### PR TITLE
Package monorobot.0.1

### PR DESCRIPTION
**Pull request by yasunariw**
### `monorobot.0.1`
Notification bot for monorepos
Notification bot to handle webhook events and post notifications to Slack.

---
* Homepage: https://github.com/ahrefs/monorobot
* Source repo: git+https://github.com/ahrefs/monorobot.git
* Bug tracker: https://github.com/ahrefs/monorobot/issues

---
:camel: Pull-request generated by opam-publish v2.1.0

### CODE REVIEW

File: packages/monorobot/monorobot.0.1/opam [hunk 1]
```diff
@@ -0,0 +1,37 @@
+opam-version: "2.0"
+synopsis: "Notification bot for monorepos"
```
Comment (kit-ty-kate) [inline packages/monorobot/monorobot.0.1/opam:2]:
```suggestion
license: "<insert your license here>"
synopsis: "Notification bot for monorepos"
```

File: packages/monorobot/monorobot.0.1/opam [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ "lwt_ppx"
+ "nocrypto"
+ "re2"
+ "stdio"
+ "uri"
+ "ocamlformat" {dev}
+ "omd" {<= "1.3.1"}
+ "yojson"
+]
+build: ["dune" "build" "-p" name]
```
Comment (kit-ty-kate) [inline packages/monorobot/monorobot.0.1/opam:30]:
```suggestion
build: ["dune" "build" "-p" name "-j" jobs]
```
Without it, setting the maxiumum amount of jobs in opam would not work.

**Review by kit-ty-kate — CHANGES_REQUESTED** [review]
Could you add the missing lower-bound constraints:
```
#=== ERROR while compiling monorobot.0.1 ======================================#
# context              2.1.1 | linux/x86_64 | ocaml-base-compiler.4.03.0 | pinned(https://github.com/ahrefs/monorobot/archive/0.1.tar.gz)
# path                 ~/.opam/4.03/.opam-switch/build/monorobot.0.1
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p monorobot
# exit-code            1
# env-file             ~/.opam/log/monorobot-35658-00e5e2.env
# output-file          ~/.opam/log/monorobot-35658-00e5e2.out
### output ###
#  gen_version src/version.ml
# fatal: not a git repository (or any parent up to mount point /)
# Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
# File "src/dune", line 2, characters 47-58:
# 2 |  (libraries lib base base.caml cmdliner devkit devkit.core extlib lwt.unix
#                                                    ^^^^^^^^^^^
# Error: Library "devkit.core" not found.
# Hint: try:
#   dune external-lib-deps --missing --no-config --root . --ignore-promoted-rules --default-target @install --always-show-command-line --promote-install-files --release --only-packages monorobot -p monorobot --profile release @install

- File "src/dune", line 2, characters 47-58:
- 2 |  (libraries lib base base.caml cmdliner devkit devkit.core extlib lwt.unix
-                                                    ^^^^^^^^^^^
- Error: Library "devkit.core" not found.
- Hint: try:
-   dune external-lib-deps --missing --no-config --root . --ignore-promoted-rules --default-target @install --always-show-command-line --promote-install-files --release --only-packages monorobot -p monorobot --profile release @install
```
```
#=== ERROR while compiling monorobot.0.1 ======================================#
# context              2.1.1 | linux/x86_64 | ocaml-base-compiler.4.05.0 | pinned(https://github.com/ahrefs/monorobot/archive/0.1.tar.gz)
# path                 ~/.opam/4.05/.opam-switch/build/monorobot.0.1
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p monorobot
# exit-code            1
# env-file             ~/.opam/log/monorobot-33735-63d443.env
# output-file          ~/.opam/log/monorobot-33735-63d443.out
### output ###
# File "/home/opam/.opam/4.05/lib/ppx_inline_test/ppx_inline_test.dune", line 1, characters 0-0:
# Warning: .dune files are ignored since 2.0. Reinstall the library with dune
# >= 2.0 to get rid of this warning and enable support for the subsystem this
# library provides.
# File "/home/opam/.opam/4.05/lib/ppx_expect/ppx_expect.dune", line 1, characters 0-0:
# Warning: .dune files are ignored since 2.0. Reinstall the library with dune
# >= 2.0 to get rid of this warning and enable support for the subsystem this
# library provides.
#  gen_version src/version.ml
# fatal: not a git repository (or any parent up to mount point /)
# Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
#       ocamlc lib/.lib.objs/byte/lib__Atd_adapters.{cmi,cmo,cmt} (exit 2)
# (cd _build/default && /home/opam/.opam/4.05/bin/ocamlc.opt -w -40 -g -bin-annot -I lib/.lib.objs/byte -I /home/opam/.opam/4.05/lib/atdgen -I /home/opam/.opam/4.05/lib/atdgen-runtime -I /home/opam/.opam/4.05/lib/base -I /home/opam/.opam/4.05/lib/base/caml -I /home/opam/.opam/4.05/lib/base64 -I /home/opam/.opam/4.05/lib/biniou -I /home/opam/.opam/4.05/lib/cstruct -I /home/opam/.opam/4.05/lib/curl -I /home/opam/.opam/4.05/lib/devkit -I /home/opam/.opam/4.05/lib/devkit/core -I /home/opam/.opam/4.05/lib/extlib -I /home/opam/.opam/4.05/lib/hex -I /home/opam/.opam/4.05/lib/lwt -I /home/opam/.opam/4.05/lib/lwt/unix -I /home/opam/.opam/4.05/lib/nocrypto -I /home/opam/.opam/4.05/lib/omd -I /home/opam/.opam/4.05/lib/re2 -I /home/opam/.opam/4.05/lib/stdio -I /home/opam/.opam/4.05/lib/uri -I /home/opam/.opam/4.05/lib/yojson -no-alias-deps -open Lib -o lib/.lib.objs/byte/lib__Atd_adapters.cmo -c -impl lib/atd_adapters.pp.ml)
# File "lib/atd_adapters.ml", line 26, characters 16-37:
# Error: Exception patterns must be at the top level of a match case.
#       ocamlc lib/.lib.objs/byte/lib__Github.{cmi,cmo,cmt} (exit 2)
# (cd _build/default && /home/opam/.opam/4.05/bin/ocamlc.opt -w -40 -g -bin-annot -I lib/.lib.objs/byte -I /home/opam/.opam/4.05/lib/atdgen -I /home/opam/.opam/4.05/lib/atdgen-runtime -I /home/opam/.opam/4.05/lib/base -I /home/opam/.opam/4.05/lib/base/caml -I /home/opam/.opam/4.05/lib/base64 -I /home/opam/.opam/4.05/lib/biniou -I /home/opam/.opam/4.05/lib/cstruct -I /home/opam/.opam/4.05/lib/curl -I /home/opam/.opam/4.05/lib/devkit -I /home/opam/.opam/4.05/lib/devkit/core -I /home/opam/.opam/4.05/lib/extlib -I /home/opam/.opam/4.05/lib/hex -I /home/opam/.opam/4.05/lib/lwt -I /home/opam/.opam/4.05/lib/lwt/unix -I /home/opam/.opam/4.05/lib/nocrypto -I /home/opam/.opam/4.05/lib/omd -I /home/opam/.opam/4.05/lib/re2 -I /home/opam/.opam/4.05/lib/stdio -I /home/opam/.opam/4.05/lib/uri -I /home/opam/.opam/4.05/lib/yojson -no-alias-deps -open Lib -o lib/.lib.objs/byte/lib__Github.cmo -c -impl lib/github.pp.ml)
# File "lib/github.ml", line 139, characters 8-55:
# Error: Exception patterns must be at the top level of a match case.
```
```
#=== ERROR while compiling monorobot.0.1 ======================================#
# context              2.1.1 | linux/x86_64 | ocaml-base-compiler.4.11.2 | pinned(https://github.com/ahrefs/monorobot/archive/0.1.tar.gz)
# path                 ~/.opam/4.11/.opam-switch/build/monorobot.0.1
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p monorobot
# exit-code            1
# env-file             ~/.opam/log/monorobot-37918-32441b.env
# output-file          ~/.opam/log/monorobot-37918-32441b.out
### output ###
#  gen_version src/version.ml
# fatal: not a git repository (or any parent up to mount point /)
# Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
# File "src/dune", line 5, characters 7-14:
# 5 |   (pps lwt_ppx))
#            ^^^^^^^
# Error: Ppx dependency on a non-ppx library "lwt_ppx". If "lwt_ppx" is in fact
# a ppx rewriter library, it should have (kind ppx_rewriter) in its dune file.

- File "src/dune", line 5, characters 7-14:
- 5 |   (pps lwt_ppx))
-            ^^^^^^^
- Error: Ppx dependency on a non-ppx library "lwt_ppx". If "lwt_ppx" is in fact
- a ppx rewriter library, it should have (kind ppx_rewriter) in its dune file.
```
```
#=== ERROR while compiling monorobot.0.1 ======================================#
# context              2.1.1 | linux/x86_64 | ocaml-base-compiler.4.13.1 | pinned(https://github.com/ahrefs/monorobot/archive/0.1.tar.gz)
# path                 ~/.opam/4.13/.opam-switch/build/monorobot.0.1
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p monorobot
# exit-code            1
# env-file             ~/.opam/log/monorobot-36586-73323a.env
# output-file          ~/.opam/log/monorobot-36586-73323a.out
### output ###
# File "lib/dune", line 3, characters 19-33:
# 3 |  (libraries atdgen atdgen-runtime base base64 base.caml biniou cstruct curl curl.lwt
#                        ^^^^^^^^^^^^^^
# Error: Library "atdgen-runtime" not found.
# Hint: try:
#   dune external-lib-deps --missing -p monorobot @install
```

File: packages/monorobot/monorobot.0.1/opam [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+ "cstruct"
+ "devkit"
+ "hex"
+ "lwt"
+ "lwt_ppx"
+ "nocrypto"
+ "re2"
+ "stdio"
+ "uri"
+ "ocamlformat" {dev}
```
Comment (kit-ty-kate) [inline packages/monorobot/monorobot.0.1/opam:26]:
```suggestion
```
I’m not seeing any use of ocamlformat during the compilation. If what you want is to automatically install what you use personally you can always have a custom makefile rule or similar other script to install everything you use.

File: packages/monorobot/monorobot.0.1/opam [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+bug-reports: "https://github.com/ahrefs/monorobot/issues"
+dev-repo: "git+https://github.com/ahrefs/monorobot.git"
+depends: [
+ "ocaml" {>= "4.08.0"}
+ "dune" {>= "2.5.0"}
+ "atd" {>= "2.2.1"}
+ "atdgen" {>= "2.2.1"}
+ "atdgen-runtime" {>= "2.2.1"}
+ "base" {>= "v0.12.0"}
+ "base64"
```
Comment (mseri) [inline packages/monorobot/monorobot.0.1/opam:17]:
```suggestion
 "base64" {>= "3.0.0"}
```

This is where the Base64 module was introduced (code uses Base64.decode_exn)

File: packages/monorobot/monorobot.0.1/opam [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+ "atdgen-runtime" {>= "2.2.1"}
+ "base" {>= "v0.12.0"}
+ "base64" {>= "3.0.0"}
+ "biniou"
+ "cmdliner"
+ "cstruct"
+ "devkit" {>= "1.20210120"}
+ "extlib" {>= "1.7.8"}
+ "hex"
+ "lwt"
```
Comment (mseri) [inline packages/monorobot/monorobot.0.1/opam:24]:
```suggestion
 "lwt" {>= "5.1.0"}
```

Code uses Lwt.all

File: packages/monorobot/monorobot.0.1/opam [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+homepage: "https://github.com/ahrefs/monorobot"
+bug-reports: "https://github.com/ahrefs/monorobot/issues"
+dev-repo: "git+https://github.com/ahrefs/monorobot.git"
+depends: [
+ "ocaml" {>= "4.08.0"}
+ "dune" {>= "2.5.0"}
+ "atd" {>= "2.2.1"}
+ "atdgen" {>= "2.2.1"}
+ "atdgen-runtime" {>= "2.2.1"}
+ "base" {>= "v0.12.0"}
```
Comment (mseri) [inline packages/monorobot/monorobot.0.1/opam:16]:
```suggestion
 "base" {>= "v0.13.0"}
```

Uses List.drop_last_exn

**Comment by mseri** [comment]
I am not sure what lower bound you need for
```
#=== ERROR while compiling monorobot.0.1 ======================================#
# context              2.1.1 | linux/x86_64 | ocaml-base-compiler.4.10.2 | pinned(https://github.com/ahrefs/monorobot/archive/0.1.tar.gz)
# path                 ~/.opam/4.10/.opam-switch/build/monorobot.0.1
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p monorobot -j 31
# exit-code            1
# env-file             ~/.opam/log/monorobot-36416-4ad04e.env
# output-file          ~/.opam/log/monorobot-36416-4ad04e.out
### output ###

[...]
# File "/home/opam/.opam/4.10/lib/ppx_derivers/ppx_derivers.dune", line 1, characters 0-0:
# Warning: .dune files are ignored since 2.0. Reinstall the library with dune
# >= 2.0 to get rid of this warning and enable support for the subsystem this
# library provides.
#  gen_version src/version.ml
# fatal: not a git repository (or any parent up to mount point /)
# Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
#       ocamlc src/.monorobot.eobjs/byte/dune__exe__Monorobot.{cmi,cmo,cmt} (exit 2)
# (cd _build/default && /home/opam/.opam/4.10/bin/ocamlc.opt -w -40 -g -bin-annot -I src/.monorobot.eobjs/byte -I /home/opam/.opam/4.10/lib/base -I /home/opam/.opam/4.10/lib/base/caml -I /home/opam/.opam/4.10/lib/cmdliner -I /home/opam/.opam/4.10/lib/devkit -I /home/opam/.opam/4.10/lib/devkit/core -I /home/opam/.opam/4.10/lib/extlib -I /home/opam/.opam/4.10/lib/lwt -I /home/opam/.opam/4.10/lib/lwt/unix -I /home/opam/.opam/4.10/lib/stdio -I /home/opam/.opam/4.10/lib/uri -I lib/.lib.objs/byte -I lib/.lib.objs/native -no-alias-deps -open Dune__exe -o src/.monorobot.eobjs/byte/dune__exe__Monorobot.cmo -c -impl src/monorobot.pp.ml)
# File "src/monorobot.ml", line 105, characters 19-43:
# 105 |   let term = Term.(const http_server_action $ addr $ port $ config $ secrets $ state) in
#                          ^^^^^^^^^^^^^^^^^^^^^^^^
# Error: This expression has type
#          unit ->
#          string ->
#          int -> Base.string -> Base.string -> Base.string option -> unit
#        but an expression was expected of type ('a -> 'b) Cmdliner.Term.t
```

**Comment by mseri** [comment]
The remaining issue is due to ocamlnet being incompatible with multicore and was fixed separately. This is good to merge.

@yasunariw please port all the fixes to your upstream development branch

**Comment by yasunariw** [comment]
Thanks @mseri @kit-ty-kate for assistance! Fixes ported upstream.

**Pull request merged**